### PR TITLE
bpo-36796: Clean the error handling in _testcapimodule.c

### PR DESCRIPTION
**Pull request by ZackerySpytz**
In addition, fix possible ref counting bugs in test_L_code() and
test_k_code() -- it was possible for "num" to be decrefed twice.

https://bugs.python.org/issue36796

### CODE REVIEW

**Comment by pganssle** [comment]
The timezone portion of the changes looks good to me, haven't looked at the other parts.

File: Modules/_testcapimodule.c [hunk 1]
```diff
@@ -851,8 +855,7 @@ test_long_as_double(PyObject *self, PyObject *Py_UNUSED(ignored))
 }
 
 /* Test the L code for PyArg_ParseTuple.  This should deliver a long long
-   for both long and int arguments.  The test may leak a little memory if
-   it fails.
+   for both long and int arguments.
```
Comment (serhiy-storchaka) [inline Modules/_testcapimodule.c]: "both long and int arguments" refers to Python types `long` and `int`. There is only one integer type in Python 3.

File: Modules/_testcapimodule.c [hunk 2]
```diff
@@ -865,36 +868,44 @@ test_L_code(PyObject *self, PyObject *Py_UNUSED(ignored))
...
-    if (value != 42)
-        return raiseTestError("test_L_code",
+    if (value != 42) {
+        raiseTestError("test_L_code",
             "L code returned wrong value for long 42");
+        goto error;
+    }
 
-    Py_DECREF(num);
     num = PyLong_FromLong(42);
```
Comment (serhiy-storchaka) [inline Modules/_testcapimodule.c]: The following code is redundant. In Python 2 there were tests for `PyInt_FromLong` and `PyLong_FromLong`.

File: Modules/_testcapimodule.c [hunk 3]
```diff
@@ -865,36 +868,44 @@ test_L_code(PyObject *self, PyObject *Py_UNUSED(ignored))
...
-    if (num == NULL)
-        return NULL;
+    if (num == NULL) {
+        goto error;
+    }
 
-    PyTuple_SET_ITEM(tuple, 0, num);
+    PyTuple_SetItem(tuple, 0, num);
```
Comment (serhiy-storchaka) [inline Modules/_testcapimodule.c]: Why it has been changed?

File: Modules/_testcapimodule.c [hunk 4]
```diff
@@ -2143,10 +2199,20 @@ static PyObject *
 test_null_strings(PyObject *self, PyObject *Py_UNUSED(ignored))
 {
     PyObject *o1 = PyObject_Str(NULL), *o2 = PyObject_Str(NULL);
+    if (o1 == NULL || o2 == NULL) {
+        goto error;
+    }
     PyObject *tuple = PyTuple_Pack(2, o1, o2);
+    if (tuple == NULL) {
```
Comment (serhiy-storchaka) [inline Modules/_testcapimodule.c]: This is not needed.

File: Modules/_testcapimodule.c [hunk 5]
```diff
@@ -2143,10 +2199,20 @@ static PyObject *
 test_null_strings(PyObject *self, PyObject *Py_UNUSED(ignored))
 {
     PyObject *o1 = PyObject_Str(NULL), *o2 = PyObject_Str(NULL);
```
Comment (serhiy-storchaka) [inline Modules/_testcapimodule.c]: In Python 2 this test tested `PyObject_Str` and `PyObject_Unicode`. In Python 3 there is no `PyObject_Unicode` and there is no need to call `PyObject_Str` twice. Just return `PyObject_Str(NULL)`.

**Comment by ZackerySpytz** [comment]
Hi, @serhiy-storchaka. Thank you for the review.

This PR deals solely with missing failure checks and possible ref counting bugs. If it is desired to make other changes to some functions in _testcapimodule.c, then perhaps that should be handled in a separate issue.

File: Modules/_testcapimodule.c (hunk 4 above)
Comment (ZackerySpytz) [reply Modules/_testcapimodule.c]: You are right.

**Comment by matrixise** [comment]
@serhiy-storchaka do you continue to review the PR of @ZackerySpytz ? Thank you

**Review by serhiy-storchaka — APPROVED** [review]

**Pull request merged**